File: Hash Table/599. Minimum Index Sum of Two Lists.cpp
/*
599. Minimum Index Sum of Two Lists

Suppose Andy and Doris want to choose a restaurant for dinner, and they both have a list of favorite restaurants represented by strings.

You need to help them find out their common interest with the least list index sum. 
If there is a choice tie between answers, output all of them with no order requirement. You could assume there always exists an answer.

Example 1:
Input:
["Shogun", "Tapioca Express", "Burger King", "KFC"]
["Piatti", "The Grill at Torrey Pines", "Hungry Hunter Steakhouse", "Shogun"]
Output: ["Shogun"]
Explanation: The only restaurant they both like is "Shogun".
Example 2:
Input:
["Shogun", "Tapioca Express", "Burger King", "KFC"]
["KFC", "Shogun", "Burger King"]
Output: ["Shogun"]
Explanation: The restaurant they both like and have the least index sum is "Shogun" with index sum 1 (0+1).
Note:
The length of both lists will be in the range of [1, 1000].
The length of strings in both lists will be in the range of [1, 30].
The index is starting from 0 to the list length minus 1.
No duplicates in both lists.

*/

class Solution {
public:
    vector<string> findRestaurant(vector<string>& list1, vector<string>& list2) {
        unordered_map<string,int>m;
        vector<string>res;
        for(int i = 0;i <list1.size();i++)
            m[list1[i]] = i;
        int minindex = list1.size()+list2.size();
        for(int i = 0; i<list2.size();i++){
            if(m.count(list2[i])>0){
                if(m[list2[i]]+i<minindex) {
                res={list2[i]};
                minindex = m[list2[i]]+i;
                }
                else if(m[list2[i]]+i==minindex) res.push_back(list2[i]);
            }

        }
        return res;
    }
};

File: Hash Table/599. Minimum Index Sum of Two Lists.py
"""
599. Minimum Index Sum of Two Lists


Example 1:
Input:
["Shogun", "Tapioca Express", "Burger King", "KFC"]
["Piatti", "The Grill at Torrey Pines", "Hungry Hunter Steakhouse", "Shogun"]
Output: ["Shogun"]
Explanation: The only restaurant they both like is "Shogun".
Example 2:
Input:
["Shogun", "Tapioca Express", "Burger King", "KFC"]
["KFC", "Shogun", "Burger King"]
Output: ["Shogun"]
Explanation: The restaurant they both like and have the least index sum is "Shogun" with index sum 1 (0+1).
"""


class Solution:
    def findRestaurant(self, list1, list2):
        """
        :type list1: List[str]
        :type list2: List[str]
        :rtype: List[str]
        """
        dic = collections.defaultdict(int)
        res, min_ind = [], float('inf')
        for i, v in enumerate(list1):
            dic[v] = i
        for i, v in enumerate(list2):
            if v in dic:
                if min_ind > i + dic[v]:
                    min_ind, res =  i + dic[v], [v]
                elif min_ind == i + dic[v]:
                    res.append(v)
        return res